extensions/amp-web-push/0.1/amp-web-push-widget.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import {TAG} from './vars';
+import {Layout} from '../../../src/layout';
+import {getServiceForDoc} from '../../../src/service';
+
+
+/** @enum {string} */
+export const WebPushWidgetVisibilities = {
+  /*
+   * Describes the state when the user is subscribed.
+   */
+  SUBSCRIBED: 'subscribed',
+  /*
+   * Describes the state when the user is not subscribed.
+   */
+  UNSUBSCRIBED: 'unsubscribed',
+  /*
+   * Widgets shown when the user has blocked permissions, or has tried
+   * subscribing in Incognito mode.
+   */
+  BLOCKED: 'blocked',
+};
+
+/** @enum {string} */
+export const WebPushWidgetActions = {
+  SUBSCRIBE: 'subscribe',
+  UNSUBSCRIBE: 'unsubscribe',
+};
+
+export class WebPushWidget extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout == Layout.FIXED;
+  }
+
+  /** @override */
+  buildCallback() {
+    // Hide the element
+    this.element.classList.add('invisible');
+
+    this.registerAction(WebPushWidgetActions.SUBSCRIBE,
+        this.onSubscribe.bind(this));
+    this.registerAction(WebPushWidgetActions.UNSUBSCRIBE,
+        this.onUnsubscribe.bind(this));
+  }
+
+  onSubscribe() {
```

Comment:
Went around and marked `_` (underscore) methods as private. Haven't gotten to all of them / some may change because some may become public again for testing.